### PR TITLE
12.0 add purchase invoice new picking line

### PR DESCRIPTION
**Pull request by robinkeunen**
[Réception d'un nouvel article - facture n'en tient pas compte -> à faire en v12](https://gestion.coopiteasy.be/web#id=1935&view_type=form&model=project.task&action=479)

tl;dr: This module invoices products that are in the picking and not in the PO.

In the normal purchase flow:
1. create a purchase order and validate PO
    - an incoming picking is created
2. receive product with delivered quantity and validate picking (no backorder)
    - the received quantity are reported on the PO
3. Create bill
    - the bill is created with products from the PO with invoiced quantity = received quantity

In food supermarket, when the supplier does not have a product A, he/she will replace it with a similar product B.
-> on the picking, the user sets "received quantity = 0" for A and adds a line for product B.

In Odoo standard, the product A is invoiced with quantity = 0 (correct) but product B is not invoiced.

This module invoices products that are in the picking and not in the PO.

## Once validated by Coop IT Easy, do not merge, submit to OCA.

### CODE REVIEW

File: purchase_invoice_new_picking_line/models/account_invoice.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+            "discount": 0.0,
+            # fixme not sure what to do with this
+            # "account_analytic_id": move.account_analytic_id.id,
+            # "analytic_tag_ids": move.analytic_tag_ids.ids,
+            "invoice_line_tax_ids": invoice_line_tax_ids.ids,
+        }
+
+        return data
+
+    @api.onchange("purchase_id")
```
Comment (houssine78) [inline purchase_invoice_new_picking_line/models/account_invoice.py:62]: this is a new function ? from my understanding it's only triggered when the purchase order id is modified... I don't see any use case where the purchase order id will change.

I probably miss something

File: purchase_invoice_new_picking_line/models/account_invoice.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    def purchase_order_change(self):
+        res = super(AccountInvoice, self).purchase_order_change()
+        po = self.env["purchase.order"].search([("name", "=", self.origin)])
+        moves = po.picking_ids.filtered(
+            lambda p: p.state == "done"
+        ).move_ids_without_package
```
Comment (houssine78) [inline purchase_invoice_new_picking_line/models/account_invoice.py]: maybe add a disclaimer in the readme to notify that package aren't taken into account by this module.

File: purchase_invoice_new_picking_line/models/account_invoice.py [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+        )
+        new_lines = self.env["account.invoice.line"]
+        for move in moves_to_invoice:
+            data = self._prepare_invoice_line_from_move(move)
+            new_line = new_lines.new(data)
+            new_line._set_additional_fields(self)
+            new_lines += new_line
+
+        self.invoice_line_ids += new_lines
+        return res
```
Comment (houssine78) [inline purchase_invoice_new_picking_line/models/account_invoice.py:81]: what if the function is called again ? the move will be invoiced a second time ? it seems to

**Review by houssine78 — APPROVED** [review]
LGTM I've just some questions

File: purchase_invoice_new_picking_line/models/account_invoice.py (hunk 1 above)
Comment (robinkeunen) [reply purchase_invoice_new_picking_line/models/account_invoice.py:62]: That trigger is complex but, basically, 
- on PO, on click create bill execution, the bill is created w/ supplier but without invoice lines
- once created (ie out of the `def create`), the field purchase_id is set thus triggering this function
- this _super_ function creates then the lines on the invoice.

File: purchase_invoice_new_picking_line/models/account_invoice.py (hunk 2 above)
Comment (robinkeunen) [reply purchase_invoice_new_picking_line/models/account_invoice.py]: k

File: purchase_invoice_new_picking_line/models/account_invoice.py (hunk 3 above)
Comment (robinkeunen) [reply purchase_invoice_new_picking_line/models/account_invoice.py:81]: yes it will, will change that!

**Comment by robinkeunen** [comment]
Cette PR a été testée pour nos _use cases_ et est prête à être déployée mais reste à soumettre à l'OCA. Je ne l'ai pas soumise car un usecase n'est pas couvert (j'ai laissé un peu trop trainer donc je ne suis plus sur duquel :/)

**Comment by houssine78** [comment]
@robinkeunen ok maintenant que l'on met ça en prod. il faudra plus trop tarder à la soumettre. Tu pourras ajouter dans le Readme le use case non couvert.

**Comment by houssine78** [comment]
@robinkeunen j'ai couvert le use case manquant => cas de plusieurs picking (backorder). Go for la PR à l'OCA

**Comment by robinkeunen** [comment]
haha, tu t'es chauffé, merci. Je me le note 👌

**Comment by houssine78** [comment]
c'est le bug en prod qui m'a chauffé ;)

**Comment by vvrossem** [comment]
Est-ce que cette PR a été soumise à l'OCA ?

**Comment by robinkeunen** [comment]
@vvrossem non :/

**Pull request closed without merging**